### DISCUSSION TITLE
[aws-events] API Destination Input Transformer missing quotes

### DISCUSSION THREAD

**Discussion by sethgeoghegan** [aws/aws-cdk#31451] [Q&A]
### Problem Summary
 
I have an EventBridge event that I want to send to an API Destination target.  The Input Transformer generated by CDK is not quoting strings, which is resulting in invalid JSON in the API request made by the API Destination.  

I am assuming this is not a bug, but a misunderstanding on my behalf on how the CDK constructs are meant to be used.

### Details

The EventBridge payload looks like this

```json
{
  "version": "0",
  "id": "89d1a02d-5ec7-412e-82f5-13505f849b41",
  "detail-type": "Scheduled Event",
  "source": "aws.events",
  "account": "123456789012",
  "time": "2016-12-30T18:44:49Z",
  "region": "us-east-1",
  "resources": ["arn:aws:events:us-east-1:123456789012:rule/SampleRule"],
  "detail": {
    "ended_at":"2024-08-29T04:51:12.778Z",
  }
}
```

I want to transform this event into an API request to `https://someapi.com/` with the `detail.ended_at` field passed in the body of the API request.  In pseudocode:

```
POST https://someapi.com/
Content-Type: application/json

{
  "timestamp": "2024-08-29T04:51:12.778Z"
}
```

I'm creating an API Destination and Input Transformer as follows:

```typescript
import * as targets from "aws-cdk-lib/aws-events-targets";
import * as events from "aws-cdk-lib/aws-events";

const myApiDest = new events.ApiDestination(stack, "myApiDest", {
    connection: myConnection,
    endpoint: "https://someapi.com/",
  })
  
const myTarget = new targets.ApiDestination(myApiDest, {
     event: events.RuleTargetInput.fromObject({
         timestamp: events.EventField.fromPath('$.detail.ended_at')
     })
})
```

This creates an Input Transformer that does _not_ escape the `ended_at` field with quotes, which results in invalid JSON. 

<img width="549" alt="Screenshot 2024-09-15 at 8 07 46 PM" src="https://github.com/user-attachments/assets/b45af6ca-e839-4097-9f39-6456ce4281cc">

I can manually update the Template in the AWS Console, which results in the valid JSON output I'm looking for:

<img width="539" alt="Screenshot 2024-09-15 at 8 09 44 PM" src="https://github.com/user-attachments/assets/62abd5a1-a898-4b10-ac72-13782912c3e7">

However, I _cannot_ figure out how to use CDK to generate a template that includes quotes around the value of `$.detail.ended_at`.  However, I've gotten rather close with this attempt:

```javascript
const myTarget = new targets.ApiDestination(myApiDest, {
     event: events.RuleTargetInput.fromObject({
         timestamp: `${events.EventField.fromPath('$.detail.ended_at')}`,
         timestamp2: `TEST${events.EventField.fromPath('$.detail.ended_at')}`,
     })
})
```

Which output the following.  The `timestamp` field lacks any quotes and is invalid JSON.  `timestamp2` has a `TEST` prefix, which is also not what I want.

```javascript
{
  "timestamp": 2024 - 08 - 29 T04: 51: 12.778 Z,
  "timestamp2": "TEST2024-08-29T04:51:12.778Z"
}
```

**Comment by brunorubin** [comment]
Hey @sethgeoghegan have you ever found a solution/workaround for this? I'm facing the same situation.

**Reply by sethgeoghegan** [reply]
Unfortunately, I don't believe I ever solved this problem.  I don't recall the specific project I was working on, but I have no memory of ever tracking down the _right_ way to do this.  Sorry I don't have better news!

**Comment by ama-joel-kravets** [comment]
For anyone looking for a solution to this problem, you can manually construct the input path and input template by creating a subclass of the `RuleTargetInput`. Here is an example of some code that does this inline.

```typescript
// To implement the cdk.aws_events.RuleTargetInput subclass you
// just need to implement the bind function.
var sqsTarget = new cdk.aws_events_targets.SqsQueue(queue, {     
	message: {
	  bind: (rule) => {
		return {
		  inputPathsMap: {
			prop1: "$.detail.prop1",
		  },
		  inputTemplate: JSON.stringify({
                        prop1Renamed: "<prop1>",
			prop2: "This is a static value",
		  }),
		} as cdk.aws_events.RuleTargetInputProperties;
	  },
	},
  });
```
